### PR TITLE
[HZ-1208] Fix persistence config merge from default map

### PR DESCRIPTION
**Pull request by TomaszGaweda**
Merge persistence & hot-restart when getting MapConfig; while this is missing, map configs based on default or wildcard configs will not have their `DataPersistenceConfig` & `HotRestartConfig` merged, resulting in unexpected EE-side persistence behaviour (see [HZ-1208](https://hazelcast.atlassian.net/browse/HZ-1208)).

With these fixes, it is revealed that assertions in `ConfigSearchTest` and `DynamicConfigSearchOrderTest` are faulty, since they are based on config name. The tests are now fixed to compare on other properties of the config (eg maxIdleSeconds).

EE PR: https://github.com/hazelcast/hazelcast-enterprise/pull/5336 (only adds test on EE side; by @vbekiaris)

Checklist:
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/config/Config.java [hunk 1]
```diff
@@ -1095,7 +1097,7 @@ public ListConfig findListConfig(String name) {
      *                                       found
      * @see StringPartitioningStrategy#getBaseName(java.lang.String)
      * @see #setConfigPatternMatcher(ConfigPatternMatcher)
-     * @see #getConfigPatternMatcher()
+     * @see #getConfigPatternMatcher()`
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/config/Config.java:1100]: Typo?

File: hazelcast/src/main/java/com/hazelcast/internal/config/ConfigUtils.java [hunk 2]
```diff
@@ -128,6 +126,16 @@ public static <T> T getConfig(ConfigPatternMatcher configPatternMatcher,
...
     }
 
+    @SuppressWarnings("unchecked")
+    private static <T> T cloneClass(Class<?> clazz, T defConfig) throws NoSuchMethodException, InstantiationException,
+            IllegalAccessException, InvocationTargetException {
+        T config;
+        Constructor<?> copyConstructor = clazz.getDeclaredConstructor(clazz);
+        copyConstructor.setAccessible(true);
+        config = (T) copyConstructor.newInstance(defConfig);
+        return config;
```
Comment (frant-hartm) [inline hazelcast/src/main/java/com/hazelcast/internal/config/ConfigUtils.java:136]:
```suggestion
        return copyConstructor.newInstance(defConfig);
```

**Review by frant-hartm — CHANGES_REQUESTED** [review]

File: hazelcast/src/test/java/com/hazelcast/config/ConfigTest.java [hunk 3]
```diff
@@ -72,24 +73,37 @@ public void testInheritanceFromDefaultConfig() {
...
+            .setBackupCount(5);
 
         HazelcastInstance hz = createHazelcastInstance(config);
 
         MapConfig binaryMapConfig = hz.getConfig().findMapConfig("myBinaryMap");
         assertEqualsStringFormat("Expected %d sync backups, but found %d", 3, binaryMapConfig.getBackupCount());
         assertEqualsStringFormat("Expected %s in-memory format, but found %s",
           MapConfig.DEFAULT_IN_MEMORY_FORMAT, binaryMapConfig.getInMemoryFormat());
+        assertEquals("myBinaryMap", binaryMapConfig.getName());
+        assertTrue(binaryMapConfig.getHotRestartConfig().isFsync());
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/config/ConfigTest.java:98]: and it win - fsync value is from data persistence

File: hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigSearchOrderTest.java [hunk 4]
```diff
@@ -58,45 +61,45 @@ public void setUp() {
...
+                hazelcastInstance.getConfig().getMapConfig(DYNAMIC_NAME).getMaxIdleSeconds());
     }
 
     @Test
     public void testSearchConfigOrderForDynamicWildcardAndExactConfig() {
         hazelcastInstance = createHazelcastInstance(staticHazelcastConfig);
-        hazelcastInstance.getConfig().addMapConfig(new MapConfig(DYNAMIC_WILDCARD_NAME));
-        hazelcastInstance.getConfig().addMapConfig(new MapConfig(DYNAMIC_NAME));
+        hazelcastInstance.getConfig().addMapConfig(new MapConfig(DYNAMIC_WILDCARD_NAME).setMaxIdleSeconds(DYNAMIC_MAX_IDLE));
+        hazelcastInstance.getConfig().addMapConfig(new MapConfig(DYNAMIC_NAME).setMaxIdleSeconds(DYNAMIC_MAX_IDLE));
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigSearchOrderTest.java:73]: Yes, 8abd3d84f7d89895025fbdac4739b58e31c3bc96

File: hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigSearchOrderTest.java [hunk 5]
```diff
@@ -58,45 +61,45 @@ public void setUp() {
     @Test
     public void testSearchConfigForDynamicWildcardOnlyConfig() {
         hazelcastInstance = createHazelcastInstance(staticHazelcastConfig);
-        hazelcastInstance.getConfig().addMapConfig(new MapConfig(DYNAMIC_WILDCARD_NAME));
-        assertEquals("Dynamic wildcard name should match", DYNAMIC_WILDCARD_NAME,
-                hazelcastInstance.getConfig().getMapConfig(DYNAMIC_NAME).getName());
+        hazelcastInstance.getConfig().addMapConfig(new MapConfig(DYNAMIC_WILDCARD_NAME).setMaxIdleSeconds(DYNAMIC_MAX_IDLE));
+        assertEquals("Dynamic wildcard name should match", DYNAMIC_MAX_IDLE,
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/DynamicConfigSearchOrderTest.java]: Indeed, fixed this and few other: 8abd3d84f7d89895025fbdac4739b58e31c3bc96

File: hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/ConfigSearchTest.java [hunk 6]
```diff
@@ -249,7 +260,7 @@ void addDynamicConfig(HazelcastInstance hazelcastInstance) {
             @Override
             void asserts() {
                 ListConfig dataConfig = hazelcastInstance.getConfig().findListConfig(DYNAMIC_NAME);
-                assertThat(dataConfig.getName(), equalTo(DYNAMIC_NAME));
+                assertThat(dataConfig.getMaxSize(), equalTo(3));
```
[the remark this answers is not in the document]
Comment (TomaszGaweda) [reply hazelcast/src/test/java/com/hazelcast/internal/dynamicconfig/ConfigSearchTest.java]: 8abd3d84f7d89895025fbdac4739b58e31c3bc96

File: hazelcast/src/main/java/com/hazelcast/config/Config.java (hunk 1 above)
Comment (TomaszGaweda) [reply hazelcast/src/main/java/com/hazelcast/config/Config.java]: probably failed shortcut ;P 8abd3d84f7d89895025fbdac4739b58e31c3bc96

**Review by frant-hartm — APPROVED** [review]

**Review by vbekiaris — APPROVED** [review]
thanks for dealing with this PR @TomaszGaweda

**Pull request merged**